your-toolbelt.md

```diff
@@ -0,0 +1,47 @@
+
+2. The other tests we haven't made are the Functional Tests, that are the tests to the controllers. The controllers receive incomming web requeste and eventually respond with a view.
+A simple case will be 
+
+```ruby
+it 'should be testable' do
+  get '/'
+  last_response.should be_ok
+end
+``
+#### Exercise: Knows how Capybara and drivers work
+
+1. What type of tests do we use Capybara for? What are its strengths and limitations?
+-- We use the capybara to do integration tests, by simulating how a real user interacts with the app.
+The advantages of using Capybara are:
+a. No setup necessary for rails and rack applications. Works out of the box.
+b. Intuitiva API 
+c. Switch the backend - The tests uses a simulated engine of the browser, without the overhead.
+d. Powerful synchronization - so that we never have to manually wait for the asynchronous processes to complete. 
+One limitation of capybara is that the default driver is `:rack_test`, which does not support javascript and HTTP resources are not accessible outside of your rack application. 
+
+2. Explain what a capybara driver is, and what some benefits are of the default driver.
+-- The Capybara driver is to simulate the browser browser, and the main benefit is to be lightweight.
+
+3. Name a case where a driver other than the default would be necessary, then go implement it with your overlord code.
+-- Other driver than the default may be de selenium if we want to use javascript.
+
+#### Exercise: Knows why flapping tests are destructive
+
+1. First, tell me what a flapping test is. Then, explain some ways that flapping tests are more destructive than even failing or non-existent tests.
+-- A flapping test is a test that may fail, but the next time it runs, it can pass. 
+-- The flapping tests are more destructive than a failing or a non-existing test because they force the developer to visit many more tests each time we modify our codebase, leading to decrease a productivity. For a normal failing test, we actually expect the test to fail initially, and it is just a matter of writing working code to make the test pass. Furthermore, when you experience flapping tests, you lose confidence that your test is actually doing its job: verifying a specific part of the application is working properly.
+
+2. Name some gotchas when writing tests that can cause them to become fragile, and how to fix those problems.
```

Comment:
Two other very common reasons for fragile testing are:

* Coupling between the tests and business logic in the SUT
* Testing the same thing in multiple tests. This is very common in integration tests, when testing the sign in flow. You need to be signed in to test most of the app, but you don't need to actually do the sign in every time. This will make it slower and force all tests to break, if you break the sign in flow, instead of just one or two.